File: tmpl/uf.go
package tmpl

// UnionFind is a data structure that tracks a set of elements partitioned into a number of disjoint (non-overlapping) subsets.
type UnionFind struct {
	parent, rank []int
	count        int
}

// Init uf
func (uf *UnionFind) Init(n int) {
	uf.parent = make([]int, n)
	uf.rank = make([]int, n)
	uf.count = n
	for i := 0; i < n; i++ {
		uf.parent[i] = i
		uf.rank[i] = 1
	}
}

func (uf *UnionFind) Find(x int) int {
	if x == uf.parent[x] {
		return x
	}
	// compress path
	uf.parent[x] = uf.Find(uf.parent[x])
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) {
	rootX, rootY := uf.Find(x), uf.Find(y)
	if rootX == rootY {
		return
	}
	if uf.rank[rootX] < uf.rank[rootY] {
		rootX, rootY = rootY, rootX
	}
	uf.parent[rootY] = rootX
	uf.rank[rootX] += uf.rank[rootY]
	uf.count--
}

// Count return the number of sets
func (uf *UnionFind) Count() int {
	return uf.count
}

File: uf/uf.py
class UF:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [1] * n
        self.cnt = n

    def find(self, x: int) -> int:
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, p: int, q: int) -> bool:
        x, y = self.find(p), self.find(q)
        if x == y:
            return False
        if self.rank[x] > self.rank[y]:
            x, y = y, x
        self.parent[x] = y
        self.rank[y] += self.rank[x]
        self.cnt -= 1
        return True

    def count(self):
        return self.cnt
